packages/analytics-browser-test/test/helpers.ts

```diff
@@ -0,0 +1,139 @@
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+import { BaseEvent, Campaign } from '@amplitude/analytics-types';
+import { BASE_CAMPAIGN } from '@amplitude/analytics-client-common';
+import { uuidPattern } from './constants';
+
+const uuid: string = expect.stringMatching(uuidPattern) as string;
+const userAgent = expect.any(String) as string;
+const library = expect.stringMatching(/^amplitude-ts\/.+/) as string;
+const number = expect.any(Number) as number;
+
+const parseQueryString = (url: string) => {
+  const params: { [key: string]: string } = {};
+  // Extract the query string part from the URL
+  const queryString = url.split('?')[1];
+  if (queryString) {
+    // Split the query string into key-value pairs
+    queryString.split('&').forEach((part) => {
+      const [key, value] = part.split('=');
+      // Decode URI components and add to the params object
+      params[decodeURIComponent(key)] = decodeURIComponent(value);
+    });
+  }
+  return params;
+};
+
+const generateAttributionUserProps = (campaignURL: string) => {
+  const campaign = parseQueryString(campaignURL);
+  const campaignObject: Campaign = {
+    ...BASE_CAMPAIGN,
+    ...campaign,
+  };
+  const hasNoCampaign = Object.entries(campaign).length === 0;
+  const hasAllCampaign = Object.entries(campaign).length === Object.entries(BASE_CAMPAIGN).length;
+  const operationObject: { [key: string]: { [key: string]: string } } = {
+    $setOnce: {},
+    ...(!hasAllCampaign ? { $unset: {} } : {}),
+    ...(!hasNoCampaign ? { $set: {} } : {}),
+  };
+
+  const user_properties = Object.keys(campaignObject).reduce((operationObject, key) => {
+    if (campaignObject[key] !== undefined) {
+      operationObject.$setOnce['initial_' + key] = campaignObject[key] as string;
+      operationObject.$set[key] = campaignObject[key] as string; // Assert that the value is a string
+    } else {
+      operationObject.$setOnce['initial_' + key] = 'EMPTY';
+      operationObject.$unset[key] = '-';
+    }
+
+    return operationObject;
+  }, operationObject);
+
+  return user_properties;
+};
+
+export const generateEvent = (event_id: number, event_type: string): BaseEvent => {
+  return {
+    device_id: uuid,
+    event_id,
```

Comment:
can we add the property key like other fields here so we know its being set correctly?

Reply:
Fixed.